doc/allocation.md

```diff
@@ -0,0 +1,112 @@
+ 
+When set to a negative number, they indicate no constraints.
+
+When set to 0, they signal that no video should be forwarded for the associated endpoint, and this is never exceeded.
+
+When set to a positive number, the algorithm will attempt to select a layer which satisfies the constraints. If no
+layers satisfy the constraints, and there is sufficient bandwidth, the algorithm will exceed the constraints and 
+select the lowest layer. In practice this is relevant only in the case where a sender does not use simulcast (or SVC),
+and encodes a single high-resolution stream. Given enough bandwidth, the stream will be forwarded even when the receiver
+signaled low constraints.
+
+## Implementation
+The bandwidth allocation algorithm is implemented in [BandwidthAllocator](https://github.com/jitsi/jitsi-videobridge/blob/master/jvb/src/main/java/org/jitsi/videobridge/cc/allocation/BandwidthAllocator.java).
+
+It consists of 3 phases:
+### 1. Prioritize
+This phase orders the available endpoints in the desired way. It starts with the endpoints ordered by speech activity
+(dominant speaker, followed by the previous dominant speaker, etc). Then, it moves the endpoints which are NOT sending
+video to the bottom of the list (this is actually implemented in [ConferenceSpeechActivity](https://github.com/jitsi/jitsi-videobridge/blob/master/jvb/src/main/java/org/jitsi/videobridge/ConferenceSpeechActivity.java).
+Finally, the selected endpoints are moved to the TOP of the list.
+
+TODO: Update the algorithm, to only move selected endpoint when they are sending video.
+
+### 2. Apply LastN
+This phase disables video for endpoints in the list that are not among the first `LastN`. Note that the effective 
+`LastN` value comes from the number signaled by the client, potentially also limited by [configuration](https://github.com/jitsi/jitsi-videobridge/blob/master/jvb/src/main/kotlin/org/jitsi/videobridge/JvbLastN.kt).
+This is implemented by setting the `maxHeight` constraint to 0.
+
+The resulting constraints are the "effective" constraints used by the rest of the algorithm. Once calculated, they are
+announced via an event, so that the sender-side constraints can be applied. Doing this step here, early in the process,
+allows us to do "aggressive layer suspension" (i.e. set sender-side constraints based on LastN).
+
+### 3. Allocation
+The final phase is the actual allocation.
+
+#### 3.1 Initialize potential layers
+The first step is to initialize a list of layers to consider for each endpoint. It starts with the list of all layers
+for the endpoint, and prunes which should not be considered:
+
+A) The ones with resolution and frame rate higher than the constraints
+
+B) The ones which are inactive (the sending endpoint is currently not transmitting them)
+
+C) In the case of `StageView`, layers with high resolution but insufficient frame rate. The minimum frame rate [can
```

Comment:
Oof, we opened up a whole other can of worms. Currently this does NOT apply to tile view. I added tests to demonstrate (`master` behaves the same). The tests look horrible, since the bitrates of the 360p7.5 and 180p30 layers are so close we make a lot of switches as bwe increases.

I agree that we should not consider 360p15 and 360p7.5 for tile view, and it should be pretty straight forward to do, but I'd rather leave it for a separate PR since it changes the functionality.